File: 프로그래머스/2/42584. 주식가격/주식가격.cpp
#include <bits/stdc++.h>
#define TOP st.top()
using namespace std;
using pii = pair<int, int>;

vector<int> solution(vector<int> prices) {
    int sz = prices.size();
    vector<int> ans(sz);
    stack<int> st;

    for(int i=0; i<sz; i++) {
        while(!st.empty() && prices[TOP] > prices[i]) {
            ans[TOP] = i - TOP;
            st.pop();
        }
        st.push(i);
    }
    
    while(!st.empty()) {
        ans[TOP] = sz - 1 - TOP;
        st.pop();
    }

    return ans;
}

File: 프로그래머스/2/42584. 주식가격/주식가격.py
def solution(prices):
    sz = len(prices)
    answer = [0 for _ in range(sz)]
    st = []

    for i, p in enumerate(prices):
        while st and prices[st[-1]] > p:
            top = st.pop()
            answer[top] = i - top
        st.append(i)

    while st:
        top = st.pop()
        answer[top] = sz - 1 - top
    
    return answer